src/Teams/Controllers/TestRunController.cs

```diff
@@ -0,0 +1,110 @@
+        // private readonly UserManager<ApplicationUser> _userManager;
+        private ApplicationUser _applicationUser;
+
+        public TestRunController(ITestRunRepository testRunRepository,
+            ApplicationDbContext context)
+        {
+            _testRunRepository = testRunRepository;
+            _context = context;
+            // _userManager = userManager;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> Index()
+        {
+            // _applicationUser = await _userManager.GetUserAsync(User);
+            return View(await _testRunRepository.GetAllAsync());
+        }
+
+        [HttpPost]
+        public IActionResult Create([FromBody] string userId, Guid testId)
+        {
+            var newTestRun = new TestRun(userId, testId);
+            _context.TestRuns.Add(newTestRun);
+            _context.SaveChanges();
+            return RedirectToAction("Index");
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> Run(Guid testRunId)
+        {
+            var testRun = await _testRunRepository.GetByIdAsync(testRunId);
+            if (testRun == null) return NotFound();
+            var testQuestionIds =
+                _context.TestQuestions.Where(q => q.TestId == testRun.TestId).Select(x => x.Id).ToList();
+            var answers =
+                AnswerDtoConvert(
+                    _context.Answers.Where(x => testQuestionIds.Contains(x.TestQuestionId)).ToList());
+            var testRunDto = new TestRunDto(answers,
+                _context.TestQuestions.Where(x => testQuestionIds.Contains(x.Id)).ToList());
+            return View(testRunDto);
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Run(TestRunDto testRunDto)
+        {
+            var updatedTestRun = await _testRunRepository.GetByIdAsync(testRunDto.Id);
+            if (updatedTestRun == null) return NotFound();
+            foreach (var answerDto in testRunDto.Answers)
+            {
+                if (answerDto.AnswerText == "")
+                {
+                    updatedTestRun.AddAnswer(new Answer(answerDto.TestQuestionId, answerDto.AnswerOptions));
+                }
+                else
+                {
+                    updatedTestRun.AddAnswer(new Answer(answerDto.TestQuestionId, answerDto.AnswerText));
+                }
+            }
+            updatedTestRun.Finish();
+            await SaveTestRun(updatedTestRun);
+            return View(testRunDto);
+        }
+
+        /// <summary>
+        /// This method is temporary, to be deleted after testing. It reads the context, gets the TestRun by id and returns back to caller.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<TestRun> PostRunTestMethod(Guid testRunId)
+        {
+            return await _testRunRepository.GetByIdAsync(testRunId);
+        }
+
+        private async Task<bool> SaveTestRun(TestRun testRun)
```

Comment:
You injected testRunRepository. Why do you directly using context. Move method to the repository and do not inject context

Reply:
In my method Run(), I still need access to dbContext in order to extract from it the id-s of questions and answers for specific testRun - because testRun no longer stores those. In previous iterations of this pull request, testRun had foreign keys like testQuestionIds and testAnswerIds, but due to recommendations by  Viktar, I removed those fields and now only can extract these values from context.
So here we have 2 options:
1) Add these fields back to TestRun entity
2) Keep injecting dbContext into the TestRunController for the sake of being able to extract them from it.

Reply:
Never mind, I think I will just change the TestRunDto to only include testId instead of adding answerIds to it, and than fully remove context from the controller. And we can inject the corresponding context in the View model later on.